zkvm/src/constraints.rs

```diff
@@ -5,36 +5,62 @@ use bulletproofs::{r1cs, r1cs::ConstraintSystem, PedersenGens};
 use curve25519_dalek::ristretto::CompressedRistretto;
 use curve25519_dalek::scalar::Scalar;
 use std::iter::FromIterator;
-use std::ops::{Add, Neg, Sub};
+use std::ops::{Add, Neg};
 
 use crate::encoding;
 use crate::errors::VMError;
 use crate::scalar_witness::ScalarWitness;
 
+/// Variable represents a high-level R1CS variable specified by its
+/// Pedersen commitment. In ZkVM variables are actually indices to a list
+/// of stored commitments to permit commitment reblinding (see `reblind` instruction).
 #[derive(Copy, Clone, Debug)]
 pub struct Variable {
     pub(crate) index: usize,
     // the witness is located indirectly in vm::VariableCommitment
 }
 
+/// Expression is a linear combination of high-level variables (`var`),
+/// low-level variables (`alloc`) and constants.
 #[derive(Clone, Debug)]
 pub enum Expression {
+    /// Represents a constant. Operations on constants produce constants.
     Constant(ScalarWitness),
+
+    /// Linear combination of R1CS variables and constants.
     LinearCombination(Vec<(r1cs::Variable, Scalar)>, Option<ScalarWitness>),
 }
 
+/// Constraint is a boolean function of expressions and other constraints.
+/// Constraints can be evaluated to true or false. The `verify` instruction
+/// enforces that the final composition evaluates to `true` in zero knowledge.
 #[derive(Clone, Debug)]
 pub enum Constraint {
+    /// Equality constraint between two expressions.
+    /// Created by `eq` instruction.
     Eq(Expression, Expression),
+
+    /// Conjunction of two constraints: each must evaluate to true.
+    /// Created by `and` instruction.
     And(Box<Constraint>, Box<Constraint>),
+
+    /// Disjunction of two constraints: at least one must evaluate to true.
+    /// Created by `or` instruction.
     Or(Box<Constraint>, Box<Constraint>),
+
+    // TBD: add `Not(Box<Constraint>)`.
```

Comment:
I think this gave a `cargo fmt` error? from travis

Reply:
fixed, thx